File: dotnet/DiningPhilosophers/Fork.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

namespace DiningPhilosophers
{
    class Fork
    {
        public bool InUse { get; set; }

        private readonly object _lockObj = new object();

        public void Take()
        {
            lock (_lockObj)
            {
                if (InUse)
                {
                    throw new Exception("Taking fork that's in use");
                }
                InUse = true;
            }
        }

        public void Release()
        {
            InUse = false;
        }
    }
}

File: go/fork.go
package main

import (
	"sync"
)

type Fork struct {
	InUse bool
	mux   sync.Mutex
}

func (fork *Fork) Take() {
	fork.mux.Lock()
	defer fork.mux.Unlock()
	if fork.InUse {
		panic("Taking fork that's in use")
	}
	fork.InUse = true
}

func (fork *Fork) Release() {
	fork.InUse = false
}
